storage/downloader.go

```diff
@@ -0,0 +1,176 @@
+// Package storage implements the storage downloader functionality.
+package storage
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/bsv-blockchain/go-sdk/overlay/lookup"
+	crypto "github.com/bsv-blockchain/go-sdk/primitives/hash"
+	"github.com/bsv-blockchain/go-sdk/transaction"
+	"github.com/bsv-blockchain/go-sdk/transaction/template/pushdrop"
+	"github.com/bsv-blockchain/go-sdk/util"
+)
+
+// StorageDownloader handles resolving and downloading files via UHRP URLs.
+type StorageDownloader struct {
+	resolver *lookup.LookupResolver
+}
+
+// NewStorageDownloader creates a new StorageDownloader with the given config.
+func NewStorageDownloader(cfg DownloaderConfig) *StorageDownloader {
+	resolver := lookup.NewLookupResolver(&lookup.LookupResolver{
+		NetworkPreset: cfg.Network,
+	})
+	return &StorageDownloader{resolver: resolver}
+}
+
+// Resolve fetches host URLs for the given UHRP URL by querying lookup services.
+func (d *StorageDownloader) Resolve(ctx context.Context, uhrpURL string) ([]string, error) {
+	// Create a lookup question
+	queryData, err := json.Marshal(map[string]string{"uhrpUrl": uhrpURL})
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal query: %w", err)
+	}
+
+	q := &lookup.LookupQuestion{
+		Service: "ls_uhrp",
+		Query:   queryData,
+	}
+
+	// Set a timeout for the lookup
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, time.Second*5)
+	defer cancel()
+
+	// Query the lookup service
+	ans, err := d.resolver.Query(ctxWithTimeout, q, 5*time.Second)
```

Comment:
Interesting that `resolver.Query()` takes a context and a timeout. Seems like only one of the timeouts is needed, probably should be updated to only take in a context.